### PR TITLE
New speedometer format

### PR DESCRIPTION
**Pull request by braem**
Closes https://github.com/momentum-mod/game/issues/1906
Closes https://github.com/momentum-mod/game/issues/1907
Closes https://github.com/momentum-mod/game/issues/1787
Closes https://github.com/momentum-mod/game/issues/1770

Unblocks https://github.com/momentum-mod/game/issues/1767

**Functionality rework**
We no longer have 1 speedometer for a set type. Rather, players can add multiple speedometers of a type (the overall velocity one and the event ones), give it a custom name, and control which axes they care about.
There has also been some consolidation, in that the energy speedometer and different units have been removed. Ramp board and exit have been consolidated into 1 speedometer type.

**New data format**
New kv3 format allows for much easier data parsing (arrays! in 2023! can you believe it!).
We also no longer need to track speedometer order, rather it is handled implicitly on reordering arrays.

**Localizable presets**
Both speedometer and range color profile preset names are now localized. Woo!

### Checks

-   [x] **I have thoroughly tested all of the code I have modified/added/removed to ensure something else did not break**
-   [x] I have followed [semantic commit messages](https://gist.github.com/joshbuchea/6f47e86d2510bce28f8e7f42ae84c716) e.g. `feat: Add foo`, `chore: Update bar`, etc...
-   [x] My branch has a clear history of changes that can be easy to follow when being reviewed commit-by-commit
-   [x] My branch is functionally complete; the only changes to be done will be those potentially requested in code review
-   [x] All changes requested in review have been `fixup`ed into my original commits.
-   [x] Fully tokenized all my strings (no hardcoded English strings!!) and supplied [bulk JSON strings](https://docs.momentum-mod.org/guide/localization/#bulk-adding-terms) below

### CODE REVIEW

File: scripts/pages/settings/speedometer.js [hunk 1]
```diff
@@ -674,21 +644,30 @@ class RangeColorProfiles {
...
+		for (const [profileName, profileObject] of this.profileObjectMap) {
+			const profileDataKV = [];
+			profileObject.saveToKV(profileDataKV);
+
+			const profile = {};
+			profile[RangeColorProfileKeys.PROFILE_NAME] = profileName;
+			profile[RangeColorProfileKeys.PROFILE_RANGE_DATA] = profileDataKV;
+			keyvalues.push(profile);
+
 			profileObject.markAsUnmodified();
```
Comment (tsa96) [inline scripts/pages/settings/speedometer.js:668]: Another case for `.map`. Again if you make `saveToKV` return the array or object it's manipulating I think you can shorten this a great deal (I'm not sure this is exactly right):
```js
const keyvalues = this.profileObjectMap.map(([profileName, profileObject]) => {
  profileObject.markAsUnmodified();

  return {
    [RangeColorProfileKeys.PROFILE_NAME]: profileName,
    [RangeColorProfileKeys.PROFILE_RANGE_DATA]: profileObject.saveToKV([])
  };
});
```

File: scripts/pages/settings/speedometer.js [hunk 2]
```diff
@@ -716,7 +695,7 @@ class RangeColorProfiles {
 
 	static makeColorProfileNamePopup(prefilledText, OKBtnText, callback) {
 		const profileNames = [];
-		for (const profileName of Object.keys(RangeColorProfiles.objectList)) profileNames.push(profileName);
+		for (const [profileName, _] of RangeColorProfiles.profileObjectMap) profileNames.push(profileName);
```
Comment (tsa96) [inline scripts/pages/settings/speedometer.js:698]: Another `.map` case! Also can't you just leave out the underscore here? i.e. `for (const [profileName] of RangeColorProfiles.profileObjectMap)`

**Review by tsa96 — CHANGES_REQUESTED** [review]
Generally really good, I mostly just have some JS simplifications you can do.

A few nits, please can you use extra newlines between methods more, also having some very basic description of each class in the settings section would be nice, I'm just finding that file a bit dense and hard to read. Understand-ability of everything would be HUGELY improved with Typescript though, it's so much easier to follow when the data structures are clearer.

File: scripts/pages/settings/speedometer.js [hunk 3]
```diff
@@ -327,19 +313,26 @@ class RangeColorObject {
...
 	static convertCSSToKV(color) {
 		const splitColor = color.replace(/[^\d,|]/g, '').split(',');
-		return `${splitColor[0]} ${splitColor[1]} ${splitColor[2]} ${splitColor[3] * 255}`;
+		return [
+			Number.parseInt(splitColor[0]),
+			Number.parseInt(splitColor[1]),
+			Number.parseInt(splitColor[2]),
+			Number.parseInt(splitColor[3]) * 255
+		];
 	}
```
[the remark this answers is not in the document]
Comment (braem) [reply scripts/pages/settings/speedometer.js:322]: This case color is a string that is like `rgba(255, 255, 255, 1)`

[the remark this answers is not in the document]
Comment (braem) [reply scripts/pages/settings/speedometer.js:322]: Ohhh like on the split, duh

File: scripts/pages/settings/speedometer.js (hunk 2 above)
Comment (braem) [reply scripts/pages/settings/speedometer.js]: Realized that this is literally just `profileObjectMap.keys()` lmao

**Comment by braem** [comment]
Also replaced a lot of calls to use `this`

File: scripts/pages/settings/speedometer.js [hunk 4]
```diff
@@ -327,19 +313,26 @@ class RangeColorObject {
...
 	static convertKVToCSS(color) {
-		const splitColor = color.split(' ');
-		return `rgba(${splitColor[0]}, ${splitColor[1]}, ${splitColor[2]}, ${splitColor[3] / 255})`;
+		return `rgba(${color[0]}, ${color[1]}, ${color[2]}, ${color[3] / 255})`;
 	}
```
[the remark this answers is not in the document]
Comment (tsa96) [reply scripts/pages/settings/speedometer.js]: Wait I don't think this is right, you're taking a string `color` as arg, then splitting it. So I think you want `const [r, g, b, a] = color.split(' ');`.

File: scripts/pages/settings/speedometer.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (tsa96) [reply scripts/pages/settings/speedometer.js]: Sorry yeah I wrote my suggestion wrong. But here, you can do parseInt with a map i.e.
```js
const [r, g, b, a] = color.replace(/[^\d,|]/g, '').split(',').map(x => Number.parseInt(x));
```

File: scripts/pages/settings/speedometer.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (braem) [reply scripts/pages/settings/speedometer.js]: Before it was a string, but the kv3 format has it in an array

[the remark this answers is not in the document]
Comment (tsa96) [reply scripts/pages/settings/speedometer.js]: Oh! Gotcha, apologies.

**Review by tsa96 — APPROVED** [review]

**Review by Gocnak — APPROVED** [review]

**Pull request merged**